sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala

```diff
@@ -351,7 +351,7 @@ private[sql] class HDFSBackedStateStoreProvider extends StateStoreProvider with
   }
 
   override def close(): Unit = {
-    synchronized { loadedMaps.values.asScala.foreach(_.clear()) }
+    synchronized { loadedMaps.keySet().asScala.toSeq.foreach(loadedMaps.remove) }
```

Comment:
Should we add a comment here explaining that the underlying value needs to remain alive/cannot be cleared - since there could be other readers with shallow references ?

Reply:
+1

Btw, do we have any reason to remove entities one by one from loadedMaps? This is now the same as loadedMaps.clear(), unless you see some difference w.r.t. behavior on concurrent operation.

Reply:
Yea I think thats true - we could prob just clear `loadedMaps` entirely ? just to confirm - it won't somehow cause `clear` to be called on the values though right ?

Reply:
the implementation of treeMap.clear is
```
public void clear() {
        modCount++;
        size = 0;
        root = null;
    }
```
where
```
private transient Entry<K,V> root;
```
so i think it just resets for the `root` and won't clear the values

Reply:
Yes, shouldn't happen, because there is no context caller has intended to do so.

Reply:
> so i think it just resets for the root and won't clear the values

If we have belief of the implementation of GC then this shouldn't matter at all. the entry instance referenced by root is dereferenced, losing reference count, and will be GC-ed. cascading effect would apply to the whole tree and effectively all of them will be GC-ed.

It might be still a best effort of freeing memory from these entries earlier, likewise we clear the value previously, but in theory they are semantically same. I'm OK with keeping best effort of it.